### PR TITLE
Replace create_temp_directory with the new create_temporary_directory

### PR DESCRIPTION
**Pull request by MichaelOrlov**
- This PR replaces #197 and relates to this comment: https://github.com/ros2/rcpputils/pull/196#discussion_r1671225326
- The newly added `create_temporary_directory(..)` uses std::filesystem::path and doesn't have platform-specific code.
- Also deprecate `create_temp_directory(..)` and `temp_directory_path`

### CODE REVIEW

**Comment by MichaelOrlov** [comment]
cc: @ahcorde, @clalancette - ready for review.

File: test/test_filesystem_helper.cpp [hunk 1]
```diff
@@ -477,16 +480,23 @@ TEST(TestFilesystemHelper, create_temp_directory)
...
   {
     if (is_win32) {
-      EXPECT_THROW(rcpputils::fs::create_temp_directory("illegalchar?"), std::system_error);
+      EXPECT_THROW(rcpputils::fs::create_temporary_directory("illegalchar?"), std::system_error);
     } else {
-      EXPECT_THROW(rcpputils::fs::create_temp_directory("base/name"), std::system_error);
+      std::filesystem::path tmpdir_with_slash_in_base;
+      EXPECT_NO_THROW(
+        tmpdir_with_slash_in_base = rcpputils::fs::create_temporary_directory("base/name"));
+      EXPECT_NO_THROW(std::filesystem::remove_all(tmpdir_with_slash_in_base));
```
Comment (MichaelOrlov) [inline test/test_filesystem_helper.cpp:488]: Note: In the new implementation, slashes in the base name is valid case and will be treated as path delimiters when creating folders.

File: test/test_filesystem_helper.cpp [hunk 2]
```diff
@@ -495,15 +505,14 @@ TEST(TestFilesystemHelper, create_temp_directory)
   // edge case inputs
   {
     // Provided no base name we should still get a path with the 6 unique template chars
-    const auto tmpdir_emptybase = rcpputils::fs::create_temp_directory("");
+    const auto tmpdir_emptybase = rcpputils::fs::create_temporary_directory("");
     EXPECT_EQ(tmpdir_emptybase.filename().string().size(), 6u);
 
-    // Empty path doesn't exist and cannot be created
-    EXPECT_THROW(rcpputils::fs::create_temp_directory("basename", path()), std::system_error);
-
```
Comment (MichaelOrlov) [inline test/test_filesystem_helper.cpp:503]: Note: The empty path is a valid case for the new implementation. The temp folder will be created in the current folder. 
However, that is not good for tests when creating a new temporary folder from where the tests running. It could be a read-only volume or require special privileges for write access.

File: test/test_filesystem_helper.cpp (hunk 1 above)
Comment (clalancette) [reply test/test_filesystem_helper.cpp:488]: I'm not a huge fan of this behavior.  The base name is really the "template" for the name, and as such, it shouldn't be involved in creating additional subdirectories.  That seems like surprising behavior to me, and different than e.g. `mkdtemp` on Linux (which fails if the basename has a `/` in it).  So I'd suggest that if `base_name` contains a directory separator, we throw an exception.

File: src/filesystem_helper.cpp [hunk 3]
```diff
@@ -341,6 +342,33 @@ path create_temp_directory(const std::string & base_name, const path & parent_pa
...
+  while (true) {
+    snprintf(random_suffix_str, kSuffixLength, "%06x", distribution(random_generator));
+    const std::string random_dir_name = base_name + random_suffix_str;
+    path_to_temp_dir = parent_path / random_dir_name;
+    // true if the directory was newly created.
+    if (std::filesystem::create_directories(path_to_temp_dir)) {
+      break;
+    }
+    if (current_iteration == max_tries) {
+      throw std::runtime_error(
```
Comment (clalancette) [inline src/filesystem_helper.cpp:364]: Nit: we need to `#include <stdexcept>` for `std::runtime_error`.

**Review by clalancette — CHANGES_REQUESTED** [review]
Overall, looks good.  I've left two things inline to fix.

This will also have to be paired with PRs to `rcl_logging` and `rosbag2` so that they use the non-deprecated API.

File: test/test_filesystem_helper.cpp (hunk 1 above)
Comment (MichaelOrlov) [reply test/test_filesystem_helper.cpp]: Ok, I added a throwing std::invalid_argument exception if base_name contains a directory separator.

File: src/filesystem_helper.cpp (hunk 3 above)
Comment (MichaelOrlov) [reply src/filesystem_helper.cpp:370]: Added.

**Comment by MichaelOrlov** [comment]
@clalancette, I addressed your comments.

File: src/filesystem_helper.cpp [hunk 4]
```diff
@@ -341,6 +343,38 @@ path create_temp_directory(const std::string & base_name, const path & parent_pa
   return final_path;
 }
 
+std::filesystem::path create_temporary_directory(
```
Comment (ahcorde) [inline src/filesystem_helper.cpp:346]: #include `<filesystem>`

**Review by ahcorde — CHANGES_REQUESTED** [review]
Just a small nit, otherwise LGTM

File: src/filesystem_helper.cpp [hunk 5]
```diff
@@ -341,6 +343,38 @@ path create_temp_directory(const std::string & base_name, const path & parent_pa
...
+  const std::string & base_name, const std::filesystem::path & parent_path, size_t max_tries)
+{
+  auto it =
+    std::find(base_name.begin(), base_name.end(), std::filesystem::path::preferred_separator);
+  if (it != base_name.end()) {
```
Comment (clalancette) [inline src/filesystem_helper.cpp:351]: I think we can make this simpler by using the `string.find` method:

```suggestion
  if (base_name.find(std::filesystem::path::preferred_separator) != base_name.npos) {
```

File: src/filesystem_helper.cpp [hunk 6]
```diff
@@ -341,6 +343,38 @@ path create_temp_directory(const std::string & base_name, const path & parent_pa
...
+    throw std::invalid_argument("The base_name contain directory-separator");
+  }
+  // mersenne twister random generator engine seeded with the std::random_device
+  std::mt19937 random_generator(std::random_device{}());
+  std::uniform_int_distribution<> distribution(0, 999999);
```
Comment (clalancette) [inline src/filesystem_helper.cpp:356]: I think this interval is too small, now that I look at it.  When I ran some tests locally, I always got `0` as the first digit.  And that makes sense, because 999999 == F423F hex, so there will never be something in the first digit.  For 6 digits of hex, we should make this 0xFFFFFF:

```suggestion
  std::uniform_int_distribution<> distribution(0, 0xFFFFFF);
```

**Review by clalancette — CHANGES_REQUESTED** [review]

File: src/filesystem_helper.cpp (hunk 4 above)
Comment (MichaelOrlov) [reply src/filesystem_helper.cpp:346]: I've added it to the relevant header file. Therefore no need to add it to the cpp file.

File: src/filesystem_helper.cpp (hunk 5 above)
Comment (MichaelOrlov) [reply src/filesystem_helper.cpp]: Fixed in the https://github.com/ros2/rcpputils/pull/198/commits/8b4087d95df458948f20f0d533c56330d5733606 commit

File: src/filesystem_helper.cpp (hunk 6 above)
Comment (MichaelOrlov) [reply src/filesystem_helper.cpp]: @clalancette Good catch! I forgot that we convert in HEX and that in hex the range should be bigger.
Fixed in the https://github.com/ros2/rcpputils/pull/198/commits/8b4087d95df458948f20f0d533c56330d5733606 commit.

**Review by ahcorde — APPROVED** [review]

**Comment by MichaelOrlov** [comment]
Pulls: ros2/rcpputils#198
Gist: https://gist.githubusercontent.com/MichaelOrlov/053726535ccac0098e0f98cd4d8aa796/raw/b89be092beb15684cb63ffbb1ffd8f9203be2c36/ros2.repos
BUILD args:  --packages-above-and-dependencies rcpputils
TEST args:  --packages-above rcpputils
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14241
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21484/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15768/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/983/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22199/)

**Comment by MichaelOrlov** [comment]
Re-run Windows CI after addressing build failures
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22213/)

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21517/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15799/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1013/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22233/)

Included https://github.com/ros2/rosbag2/pull/1740/ and https://github.com/ros2/rcl_logging/pull/117
